pkg/cloudprovider/providers/photon/photon.go

```diff
@@ -0,0 +1,550 @@
+	"io"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+	k8stypes "k8s.io/kubernetes/pkg/types"
+	"log"
+	"os/exec"
+	"strings"
+)
+
+const (
+	ProviderName = "photon"
+)
+
+// Global variable pointing to photon client
+var Esxclient *photon.Client
+var logger *log.Logger = nil
+
+// overrideIP indicates if the hostname is overriden by IP address, such as when
+// running multi-node kubernetes using docker. In this case the user should set
+// overrideIP = true in cloud config file. Default value is false.
+var overrideIP bool = false
+
+// Photon is an implementation of cloud provider Interface for Photon Controller.
+type PCCloud struct {
+	cfg *PCConfig
+	// InstanceID of the server where this PCCloud object is instantiated.
+	localInstanceID string
+	// local $HOSTNAME
+	localHostname string
+	// hostname from K8S, could be overridden
+	localK8sHostname string
+	// Photon Project ID
+	projID string
+	cloudprovider.Zone
+}
+
+type PCConfig struct {
+	Global struct {
+		CloudTarget       string `gcfg:"target"`
+		IgnoreCertificate bool   `gcfg:"ignoreCertificate"`
+		Tenant            string `gcfg:"tenant"`
+		Project           string `gcfg:"project"`
+		OverrideIP        bool   `gcfg:"overrideIP"`
+	}
+}
+
+// Disks is interface for manipulation with PhotonController Persistent Disks.
+type Disks interface {
+	// AttachDisk attaches given disk to given node. Current node
+	// is used when nodeName is empty string.
+	AttachDisk(pdID string, nodeName k8stypes.NodeName) error
+
+	// DetachDisk detaches given disk to given node. Current node
+	// is used when nodeName is empty string.
+	DetachDisk(pdID string, nodeName k8stypes.NodeName) error
+
+	// DiskIsAttached checks if a disk is attached to the given node.
+	DiskIsAttached(pdID string, nodeName k8stypes.NodeName) (bool, error)
+
+	// CreateDisk creates a new PD with given properties.
+	CreateDisk(volumeOptions *VolumeOptions) (pdID string, err error)
+
+	// DeleteDisk deletes PD.
+	DeleteDisk(pdID string) error
+}
+
+// VolumeOptions specifies capacity, tags, name and flavorID for a volume.
+type VolumeOptions struct {
+	CapacityGB int
+	Tags       map[string]string
+	Name       string
+	Flavor     string
+}
+
+func logError(msg string, err error) error {
+	s := "Photon Cloud Provider: " + msg + ". Error [" + err.Error() + "]"
+	glog.Errorf(s)
+	return fmt.Errorf(s)
+}
+
+func readConfig(config io.Reader) (PCConfig, error) {
+	if config == nil {
+		err := fmt.Errorf("cloud provider config file is missing. Please restart kubelet with --cloud-provider=photon --cloud-config=[path_to_config_file]")
+		return PCConfig{}, err
+	}
+
+	var cfg PCConfig
+	err := gcfg.ReadInto(&cfg, config)
+	return cfg, err
+}
+
+func init() {
+	cloudprovider.RegisterCloudProvider(ProviderName, func(config io.Reader) (cloudprovider.Interface, error) {
+		cfg, err := readConfig(config)
+		if err != nil {
+			return nil, logError("failed to read in cloud provider config file", err)
```

Comment:
Do you need to include the pathname to the configuration file, or is that already embedded in the existing error message returned by readConfig?